geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs

```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+
+data TaptoolsPairOverride = TaptoolsPairOverride
+  { ttpoAsset            ∷ !GYAssetClass,
+    ttpoPrecision        ∷ !Natural
+  }
+  deriving stock (Show, Generic)
+  deriving (FromJSON, ToJSON) via CustomJSON '[FieldLabelModifier '[CamelToSnake]] TaptoolsPairOverride
+
+data PricesProvidersAggregator = PPA
+  { ppaCommonCfg     ∷ !PriceCommonCfg,
+    ppaPricesCluster ∷ ![PricesProviderCfg]
+  }
+
+data PricesProviders = PP
+  { pricesAggregatorPP ∷ !PricesProvidersAggregator,
+    orderBookPP        ∷ !(Connection, DEXInfo)
+  }
+
+buildPP
+  ∷ Connection
+  → DEXInfo
+  → PriceConfig
+  → IO PricesProviders
+buildPP c dex pc =
+  PP
+   <$> ppPricesAggregator
+   <*> return (c, dex)
+ where
+  ppPricesAggregator ∷ IO PricesProvidersAggregator
+  ppPricesAggregator = case pc of
+    PCVersion1 (PriceConfigV1 {..}) → do
+      let pccNetworkId   = pcNetworkId
+          mcApiKey       = pcApiKey
+          mcDex          = pcDex
+          mcPairOverride = pcOverride
+      return PPA
+        { ppaCommonCfg = PriceCommonCfg
+          { pccCommonResolution       = CRes5m  -- placeholder, unused
+          , pccNetworkId              = pccNetworkId
+          , pccPriceDiffThreshold1    = 1
+          , pccPriceDiffThreshold2    = 1
+          , pccRespiteDelayFactor     = 0.5     -- TODO: sensible value?
+          , pccAfterExitRelaxAim1     = 0       -- placeholder, unused
+          , pccAfterExitWorseMax1     = 0       -- placeholder, unused
+          , pccAfterExitRelaxAim2     = 10      -- TODO: sensible value?
+          , pccPricesProvidersWeights = [1]
+          }
+        , ppaPricesCluster = [ MaestroPPC MaestroConfig
+            { mcApiKey             = mcApiKey
+            , mcResolutionOverride = Just pcResolution
+            , mcDex                = mcDex
+            , mcPairOverride       = mcPairOverride
+            } ]
+        }
+
+    PCVersion2 (PriceConfigV2 {..}) → do
+      let ppaCommonCfg     = pcPriceCommonCfg
+          thresHold1       = pccPriceDiffThreshold1 ppaCommonCfg
+          thresHold2       = pccPriceDiffThreshold2 ppaCommonCfg
+          ppaPricesCluster = NE.toList pcPricesProviderCfgs
+
+      when (length (pccPricesProvidersWeights ppaCommonCfg) /= length ppaPricesCluster) $
+        throwIO $ userError "Expected 'pccPricesProvidersWeights' to be of same length as 'pcPricesProviderCfgs'."
+      when (thresHold2 < thresHold1) $
+        throwIO $ userError "Expected 'pccPriceDiffThreshold2 >= pccPriceDiffThreshold1'."
+
+      return PPA
+        { ppaCommonCfg     = ppaCommonCfg,
+          ppaPricesCluster = ppaPricesCluster
+        }
+  
+data PriceIndicator = PriceMismatch1 | PriceMismatch2 | PriceUnavailable | PriceSourceFail [String] [String] Price | PriceAverage Price
+  deriving stock Show
+
+-- | Prices provider error:  `SourceUnavailable displayException pricesProvider`.
+data SourceError = SourceUnavailable String String
```

Comment:
Can have tags for providers, like `data ProviderTag = MaestroTag | TapToolsTag`, so we can avoid `String` here. This is optional.

Reply:
Decided to keep it simple for now, and use String.